## 3 
We now review k-fold cross-validation.

## (a)  Explain how k-fold cross-validation is implemented.

- k-fold cross validation is implemented by taking a training set of size n and partitioning it into k non-overlapping equal parts. If k does not get divided into n evenly then we split the k fold as equal as we can. Then the model is fit k number of times, each time leaving out a different fold as a validation set. For each of the k model fits, the left-out folds are used to calucate the validation error. The the average of the k validation errors is use to be an estimate of the test error.

## (b) What are the advantages and disadvantages of k-fold cross- validation relative to:

(i) The validation set approach?
- Requires less computation and is simplet adn easier to implement. On the other hand, the validation set approach tents to oversestimate the test error becasue it only uses half of the sample to fit the model. 

(ii) LOOCV?
- Leave-one-out cross-validation (LOOCV) exhibits lower bias compared to k-fold cross-validation as it utilizes nearly all data points, minimizing bias. However, this comes at the expense of increased variance in LOOCV due to highly correlated validation errors, resulting from models differing by only two points. In contrast, k-fold cross-validation, especially with k=5 or k=10, reduces this correlation significantly, sharing only about 60-80% of points between pairs of fits. Moreover, k-fold cross-validation is computationally less intensive, requiring only k model fits instead of n fits in LOOCV, except in cases like least squares linear or polynomial regression where LOOCV matches the computational load of a single model fit (

## 5
In Chapter 4, we used logistic regression to predict the probability of
default using income and balance on the Default data set. We will
now estimate the test error of this logistic regression model using the
validation set approach. Do not forget to set a random seed before
beginning your analysis.

In [28]:
import pandas as pd
import numpy as np
import patsy
import seaborn as sns
import matplotlib.pyplot as plt

import statsmodels.formula.api as smf
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression, LinearRegression
from sklearn.model_selection import train_test_split, cross_val_score, LeaveOneOut, KFold
from sklearn.preprocessing import OneHotEncoder, PolynomialFeatures
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.utils import resample
from scipy import stats

import statsmodels.api as sm
import statsmodels.formula.api as smf

sns.set(style="white")
%matplotlib inline

(a) Fit a logistic regression model that uses income and balance to
predict default.

In [9]:
np.random.seed(1)

df = pd.read_csv("https://raw.githubusercontent.com/jasonm/islr-exercises/master/Data/Default.csv", index_col=0)

df['default_yes'] = (df['default'] == 'Yes').astype('int')
df.head()

default student      balance        income  default_yes
1      No      No   729.526495  44361.625074            0
2      No     Yes   817.180407  12106.134700            0
3      No      No  1073.549164  31767.138947            0
4      No      No   529.250605  35704.493935            0
5      No      No   785.655883  38463.495879            0

In [3]:
X = df[['balance', 'income']]
y = df['default']
clf = LogisticRegression(penalty = "none", solver = "lbfgs")
clf.fit(X, y)
clf.coef_

C:\Users\Tara Grey\Python\Lib\site-packages\sklearn\linear_model\_logistic.py:1182: FutureWarning: `penalty='none'`has been deprecated in 1.2 and will be removed in 1.4. To keep the past behaviour, set `penalty=None`.
  warnings.warn(


array([[5.64710291e-03, 2.08089921e-05]])

(b) Using the validation set approach, estimate the test error of this
model. In order to do this, you must perform the following steps:

i. Split the sample set into a training set and a validation set.

ii. Fit a multiple logistic regression model using only the train-
ing observations.

iii. Obtain a prediction of default status for each individual in
the validation set by computing the posterior probability of
default for that individual, and classifying the individual to
the default category if the posterior probability is greater
than 0.5.

iv. Compute the validation set error, which is the fraction of
the observations in the validation set that are misclassified.

In [8]:
X_train, X_test, y_train, y_test = train_test_split(df[["balance", "income"]], df["default"],
                                                   test_size = 0.25, random_state = 312)
clf = LogisticRegression(penalty = "none", solver = "lbfgs")
clf.fit(X_train, y_train)
1 - clf.score(X_test, y_test)

C:\Users\Tara Grey\Python\Lib\site-packages\sklearn\linear_model\_logistic.py:1182: FutureWarning: `penalty='none'`has been deprecated in 1.2 and will be removed in 1.4. To keep the past behaviour, set `penalty=None`.
  warnings.warn(


0.02839999999999998

(c) Repeat the process in (b) three times, using three different splits
of the observations into a training set and a validation set. Com-
ment on the results obtained.

In [9]:
X_train, X_test, y_train, y_test = train_test_split(df[["balance", "income"]], df["default"],
                                                   test_size = 0.25, random_state = 456)
clf = LogisticRegression(penalty = "none", solver = "lbfgs")
clf.fit(X_train, y_train)
1 - clf.score(X_test, y_test)

C:\Users\Tara Grey\Python\Lib\site-packages\sklearn\linear_model\_logistic.py:1182: FutureWarning: `penalty='none'`has been deprecated in 1.2 and will be removed in 1.4. To keep the past behaviour, set `penalty=None`.
  warnings.warn(


0.03200000000000003

In [10]:
X_train, X_test, y_train, y_test = train_test_split(df[["balance", "income"]], df["default"],
                                                   test_size = 0.25, random_state = 789)
clf = LogisticRegression(penalty = "none", solver = "lbfgs")
clf.fit(X_train, y_train)
1 - clf.score(X_test, y_test)

C:\Users\Tara Grey\Python\Lib\site-packages\sklearn\linear_model\_logistic.py:1182: FutureWarning: `penalty='none'`has been deprecated in 1.2 and will be removed in 1.4. To keep the past behaviour, set `penalty=None`.
  warnings.warn(


0.034399999999999986

In [11]:
X_train, X_test, y_train, y_test = train_test_split(df[["balance", "income"]], df["default"],
                                                   test_size = 0.25, random_state = 314159)
clf = LogisticRegression(penalty = "none", solver = "lbfgs")
clf.fit(X_train, y_train)
1 - clf.score(X_test, y_test)

C:\Users\Tara Grey\Python\Lib\site-packages\sklearn\linear_model\_logistic.py:1182: FutureWarning: `penalty='none'`has been deprecated in 1.2 and will be removed in 1.4. To keep the past behaviour, set `penalty=None`.
  warnings.warn(


0.031200000000000006

In [12]:
(df["default"] != "No").mean()

0.0333

## Comment

- Across three distinct 75-25 data splits, our validation set error demonstrated remarkable consistency. The average error, in comparison to Part 2, stood at 0.0315, with errors falling within approximately 20% of each other. It's noteworthy that these test errors closely align with the error obtained by adopting a naive strategy of predicting no defaults. The average test error is merely 5% lower than the error associated with the naive approach

(d) Now consider a logistic regression model that predicts the prob-
ability of default using income, balance, and a dummy variable
for student. Estimate the test error for this model using the val-
idation set approach. Comment on whether or not including a
dummy variable for student leads to a reduction in the test error
rate

In [13]:
import warnings
warnings.filterwarnings("ignore", category=FutureWarning)

np.random.seed(24)
with_student = {}
without_student = {}

categorical_features = ["student"]
categorical_transformer = Pipeline([("onehot", OneHotEncoder(drop = "first"))])
numerical_features = ["income", "balance"]
with_student_preprocessor = ColumnTransformer([("cat", categorical_transformer, categorical_features),
                                 ("num", "passthrough", numerical_features)])
with_student_clf = Pipeline([("preprocessor", with_student_preprocessor), 
                ("classifier", LogisticRegression(penalty = "none", solver = "lbfgs"))])
without_student_preprocessor = ColumnTransformer([("num", "passthrough", numerical_features)])
without_student_clf = Pipeline([("preprocessor", without_student_preprocessor), 
                ("classifier", LogisticRegression(penalty = "none", solver = "lbfgs"))])

# Loop through 50 train-test splits to compute average difference in error rate
for i in range(50):
    # Split the data in to training and test sets
    X_train, X_test, y_train, y_test = train_test_split(df, df["default"], test_size = 0.25)
    # Fit classifier which includes student variable and compute validation set error
    with_student_clf.fit(X_train, y_train)
    with_student[i] = 1 - with_student_clf.score(X_test, y_test)
    # Fit classifier which excludes student variable and compute validation set error
    without_student_clf.fit(X_train, y_train)
    without_student[i] = 1 - without_student_clf.score(X_test, y_test)
errors = pd.DataFrame({"with_student": with_student, "without_student": without_student})
errors["difference"] = errors["with_student"] - errors["without_student"]
errors["difference"].mean()

0.003000000000000016

## Comment: 
- Looping through the 50 train-test splits adn comparing the error rates between the logistic regression model that's predicting defult using income, balance and student adn the logistic regression model that's predicting defualt without student for each split, it can be seen that including the dummy variable for student did not lead to any reduction in the test error rate. In fact, it resulted in a slight increase in the test error rate. 

## 6
We continue to consider the use of a logistic regression model to
predict the probability of default using income and balance on the
Default data set. In particular, we will now compute estimates for the
standard errors of the income and balance logistic regression coeffi-
cients in two different ways: (1) using the bootstrap, and (2) using the
standard formula for computing the standard errors in the sm.GLM()
function. Do not forget to set a random seed before beginning your
analysis

(a) Using the summarize() and sm.GLM() functions, determine the
estimated standard errors for the coefficients associated with
income and balance in a multiple logistic regression model that
uses both predictors

In [12]:
Default = pd.read_csv("https://raw.githubusercontent.com/jasonm/islr-exercises/master/Data/Default.csv", index_col=0)


In [13]:
np.random.seed(1)
x01  = sm.add_constant(Default.iloc[:, 3:5]) 
y01  = np.where(Default['default']=='No', 0, 1) 
glm1 = sm.Logit(y01, x01)
print(glm1.fit().bse)

Optimization terminated successfully.
         Current function value: 0.145834
         Iterations 7
const     0.146257
income    0.000004
dtype: float64


(b) Write a function, boot_fn() , that takes as input the Default data
set as well as an index of the observations, and that outputs
the coefficient estimates for income and balance in the multiple
logistic regression model

In [15]:
def coef(x,y):
  glm = sm.Logit(y, x)  
  return glm.fit().params
coef(x01,y01)

Optimization terminated successfully.
         Current function value: 0.145834
         Iterations 7


const    -3.094149
income   -0.000008
dtype: float64

(c) Following the bootstrap example in the lab, use your boot_fn()
function to estimate the standard errors of the logistic regression
coefficients for income and balance

In [16]:
def boot(data, n):
    x1 = np.zeros(n)
    x2 = np.zeros(n)
    
    for i in range(0, n):
        df    = data.sample(frac=1, replace=True)
        x     = sm.add_constant(df.iloc[:, 3:5])
        y     = np.where(df['default']=='No', 0, 1) 
        x1[i] = coef(x,y)[1] 
        x2[i] = coef(x,y)[2]
        
    res1 = np.std(x1)
    res2 = np.std(x2)
    print('balance se: %.8f; income se: %.8f' %(res1, res2))
    
boot(Default, 50)


Optimization terminated successfully.
         Current function value: 0.145473
         Iterations 7
Optimization terminated successfully.
         Current function value: 0.145473
         Iterations 7


IndexError: index 2 is out of bounds for axis 0 with size 2

(d) Comment on the estimated standard errors obtained using the
sm.GLM() function and using the bootstrap


- Bootstrap Standard errors are close to the glm estimates

## 9  
We will now consider the Boston housing data set, from the ISLP library

In [18]:
Boston = pd.read_csv("Boston.csv")
Boston.head()

crim    zn  indus  chas    nox     rm   age     dis  rad  tax  ptratio  \
0  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296     15.3   
1  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242     17.8   
2  0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242     17.8   
3  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222     18.7   
4  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222     18.7   

    black  lstat  medv  
0  396.90   4.98  24.0  
1  396.90   9.14  21.6  
2  392.83   4.03  34.7  
3  394.63   2.94  33.4  
4  396.90   5.33  36.2

(a) Based on this data set, provide an estimate for the population
mean of medv. Call this estimate ˆμ

In [20]:
mu_hat = np.mean(Boston.medv) 
print(mu_hat)

22.532806324110677


(b) Provide an estimate of the standard error of μ^. Interpret this result.

In [21]:
mu_hat_se = np.std(Boston.medv)/np.sqrt(len(Boston)) 
print(mu_hat_se)

0.4084569346972867


Interpetation:
- 0.408 is our estimate for the standard error of u. Since CMEDV measured in units of 1,000 usd, this estimate translates to $408. 

(c)

In [22]:
def boot(var, n):
    m = np.zeros(n)
    for i in range(0, n):
        v    = var.sample(frac=1, replace=True)
        m[i] = np.mean(v)
    res1 = np.mean(m)
    res2 = np.std(m)
    print('mu: %.2f; se: %.2f' %(res1, res2))
    return(res1, res2)

result = boot(Boston.medv, 50)

mu: 22.45; se: 0.38


This is close to (b) but less then (b)'s standard error

(d)

In [23]:

print('lowerbd:%.2f' %(result[0] - 2*result[1]))
print('upperbd:%.2f' %(result[0] + 2*result[1])) 

from scipy import stats
stats.t.interval(0.95,               # confidence level
                 df = len(Boston)-1, # degrees of freedom
                 loc = mu_hat,       # sample mean
                 scale= mu_hat_se)   # sample std dev

lowerbd:21.69
upperbd:23.21


(21.73032216040747, 23.335290487813882)

(e)

In [24]:
mu_med_hat = np.median(Boston.medv)
print(mu_med_hat)

21.2


(f)

In [25]:
def boot(var, n):
    m = np.zeros(n)
    for i in range(0, n):
        v     = var.sample(frac=1, replace=True)
        m[i]  = np.median(v)
    r = np.std(m) 
    print(r)

result = boot(Boston.medv, 50)

0.3782856063875545


Comment: 
    - Using the bootstrap, we have an estimate of 0.38 for the standard error of μ^med. Based on this bootstrap estimate, a 95% confidence interval for the median of CMEDV is approximately [21.2−2×0.38,21.2+2×0.38], or [20.44,21.96]. There is a 95% chance that this confidence interval contains the actual population median of CMEDV.

(g)

In [26]:
mu_10_hat = Boston['medv'].quantile(q=0.1)
print(mu_10_hat) 

12.75


Comment: 
- The estimated population tenth percentile of CMEDV is μ^0.1=12.75 In other words, the tenth percentile of median home values for this sample is $12,750, or only 10% of observations in this sample have a CMEDV value of 12.75 or less.

(h)

In [27]:
def boot(var, n):
    m = np.zeros(n)
    for i in range(0, n):
        v     = var.sample(frac=1, replace=True)
        m[i]  = v.quantile(q=0.1)
    r = np.std(m) 
    print(r)

result = boot(Boston.medv, 50)

0.41239422886359617


Comment:
- Using bootstrap we have an estimate of 0.412 for the standard error of μ^0.1.